massa-ledger/src/ledger.rs

```diff
@@ -184,4 +196,244 @@ impl FinalLedger {
+                            next_cursor.1 = LedgerCursorStep::Datastore(Some(*key));
+                            if data.len() as u64 > LEDGER_PART_SIZE_MESSAGE_BYTES {
+                                return Ok((data, next_cursor));
+                            }
+                        }
+                        next_cursor.1 = LedgerCursorStep::Finish;
+                    }
+                    LedgerCursorStep::Finish => {
+                        data.push(DATASTORE_END_IDENTIFIER);
+                        next_cursor.1 = LedgerCursorStep::Start;
+                        next_cursor.0 = *addr;
+                        break;
+                    }
+                }
+                if data.len() as u64 > LEDGER_PART_SIZE_MESSAGE_BYTES {
+                    return Ok((data, next_cursor));
+                }
+            }
+        }
+        Ok((data, next_cursor))
+    }
+
+    /// Set a part of the ledger
+    /// Used for bootstrap
+    /// Parameters:
+    /// * cursor: Where we stopped in the ledger
+    ///
+    /// Returns:
+    /// Nothing on success error else.
+    pub fn set_ledger_part(
+        &mut self,
+        old_cursor: Option<LedgerCursor>,
+        new_cursor: LedgerCursor,
+        data: Vec<u8>,
+    ) -> Result<(), ModelsError> {
+        let mut data = data.as_bytes();
+        let address_deserializer = AddressDeserializer::new();
+        let hash_deserializer = HashDeserializer::new();
+        let mut cursor = if let Some(old_cursor) = old_cursor {
+            old_cursor
+        } else {
+            let (rest, address) = address_deserializer.deserialize(data).map_err(|_| {
+                ModelsError::DeserializeError("Fail to deserialize address".to_string())
+            })?;
+            data = rest;
+            self.sorted_ledger
+                .entry(address)
+                .or_insert_with(LedgerEntry::default);
+            LedgerCursor(address, LedgerCursorStep::Balance)
+        };
+        let u64_deserializer =
+            U64VarIntDeserializer::new(Bound::Included(0), Bound::Included(u64::MAX));
+        let vecu8_deserializer = VecU8Deserializer::new();
+        while cursor != new_cursor {
+            // We want to make one check per loop to check that the cursor isn't finish each loop turn.
+            match cursor.1 {
+                LedgerCursorStep::Start => {
+                    let (rest, address) = address_deserializer.deserialize(data).map_err(|_| {
+                        ModelsError::DeserializeError("Fail to deserialize address".to_string())
+                    })?;
+                    data = rest;
+                    self.sorted_ledger
+                        .entry(address)
+                        .or_insert_with(LedgerEntry::default);
+                    cursor.1 = LedgerCursorStep::Balance;
+                }
+                LedgerCursorStep::Balance => {
+                    let amount_deserializer = AmountDeserializer::new();
```

Comment:
Shouldn't this deserializer be initialized at the beginning of the function like the others?

Reply:
Done.